unified_planning/plan.py

```diff
@@ -14,12 +14,24 @@
 #
 '''This module defines the different plan classes.'''
 
+
 import unified_planning
 import unified_planning.model
 from typing import Dict, Optional, Tuple, List
 from fractions import Fraction
 
 
+ALL_STATUS = list(range(0, 8))
+
+(
+SATISFIED, OPTIMAL, UNSATISFIED, SEARCH_SPACE_EXHAUSTED,
+TIMEOUT, MEMOUT, INTERNAL_ERROR, UNSUPPORTED_PROBLEM
+) = ALL_STATUS
+
+POSITIVE_OUTCOMES = frozenset([SATISFIED, OPTIMAL])
+
+NEGATIVE_OUTCOMES = frozenset([UNSATISFIED, SEARCH_SPACE_EXHAUSTED, UNSUPPORTED_PROBLEM])
```

Comment:
Should we include also TIMEOUT, MEMOUT, INTERNAL_ERROR?

Reply:
I don't think so because NEGATIVE_OUTCOMES assumes that a plan is not found. TIMEOUT, MEMOUT and INTERNAL_ERROR, at least from my interpretation of their meaning, could return a plan, still knowing that something went wrong.

Reply:
I don't agree. They are reasons why a plan is not found.
TIMEOUT for example means that a plan is not found within the time limit

Reply:
In the case of planning with explicit costs, it could be the case that a plan was found but the planner exited (e.g. timeout) before finding the optimal one. In this case, the "best plan found" remains very important.

Reply:
Do you think this is the case also for MEMOUT, INTERNAL_ERROR and SEARCH_SPACE_EXHAUSTED?

Reply:
From what I understood, SATISFIED and OPTIMAL requires the `plan` field set while the UNSATISFIED and UNSUPPORTED_PROBLEM never have the `plan` field set. In the other cases a plan might be set, right?

Reply:
Correct except for SEARCh_SPACE_EXHAUSTED which that the planner stopped because he is not capable of generating more potential solutions (but there might other that are possible, i.e., the planner is incomplete). 

In a nutshell :
-  if a plan is provided with the SEARCH_SPACE_EXHAUSTED flag, it is a weaker version of the OPTIMAL flag (without an actual guarantee on the optimality).
- if no plan is provided with the SEARCH_SPACE_EXHAUSTED flag, it is a weaker version of the UNSATISFIABLE version : the planner reports that he is not capable of finding a solution but a complete planner might find one.

I was unsure whether to include this flag. The reasoning was that (1) we want to trust the planner when it says that something is optimal/unsat because this is not something we can routinely verify, (2) if we don't include those weaker cases, then incomplete planners will have to (incorrectly) report optimal/unsat.

This is mostly related to planner incompletness so it might handle another way (as distinct boolean flag, directly in the planner description, ...)

Reply:
I agree on what you said about the case "no plan is provided with the SEARCH_SPACE_EXHAUSTED flag".
For the case "a plan is provided with the SEARCH_SPACE_EXHAUSTED flag", I think we can express the same message with the SATISFIED flag: the planner reports that the plan is the best one found so far (with no guarantee on the optimality) and it is not able to find any better plans.